examples/simple_login.rs

```diff
@@ -0,0 +1,203 @@
+
+use opaque_ke::{
+    ciphersuite::CipherSuite,
+    keypair::KeyPair,
+    opaque::{
+        ClientLogin, ClientRegistration, LoginFirstMessage, LoginSecondMessage, LoginThirdMessage,
+        RegisterFirstMessage, RegisterSecondMessage, RegisterThirdMessage, ServerLogin,
+        ServerRegistration,
+    },
+};
+
+// The ciphersuite trait allows to specify the underlying primitives
+// that will be used in the OPAQUE protocol
+#[allow(dead_code)]
+struct Default;
+impl CipherSuite for Default {
+    type Group = curve25519_dalek::ristretto::RistrettoPoint;
+    type KeyFormat = opaque_ke::keypair::X25519KeyPair;
+    type KeyExchange = opaque_ke::key_exchange::tripledh::TripleDH;
+    type Hash = sha2::Sha256;
+    type SlowHash = opaque_ke::slow_hash::NoOpHash;
+}
+
+fn get_password_file(server_kp: &opaque_ke::keypair::X25519KeyPair, password: String) -> Vec<u8> {
+    let mut client_rng = OsRng;
+    let (r1, client_state) =
+        ClientRegistration::<Default>::start(password.as_bytes(), Some(b"pepper"), &mut client_rng)
+            .unwrap();
+    let r1_bytes = r1.to_bytes();
+
+    // Client sends r1_bytes to server
+
+    let mut server_rng = OsRng;
+    let (r2, server_state) = ServerRegistration::<Default>::start(
+        RegisterFirstMessage::try_from(&r1_bytes[..]).unwrap(),
+        &mut server_rng,
+    )
+    .unwrap();
+    let r2_bytes = r2.to_bytes();
+
+    // Server sends r2_bytes to client
+
+    let (r3, _) = client_state
+        .finish(
+            RegisterSecondMessage::try_from(&r2_bytes[..]).unwrap(),
+            server_kp.public(),
+            &mut client_rng,
+        )
+        .unwrap();
+    let r3_bytes = r3.to_bytes();
+
+    // Client sends r3_bytes to server
+
+    let password_file = server_state
+        .finish(RegisterThirdMessage::try_from(&r3_bytes[..]).unwrap())
+        .unwrap();
+    password_file.to_bytes()
+}
+
+fn check_login(
+    server_kp: &opaque_ke::keypair::X25519KeyPair,
+    password: String,
+    password_file_bytes: &[u8],
+) -> bool {
+    let mut client_rng = OsRng;
+    let (l1, client_state) =
+        ClientLogin::<Default>::start(password.as_bytes(), Some(b"pepper"), &mut client_rng)
+            .unwrap();
+    let l1_bytes = l1.to_bytes();
+
+    // Client sends l1_bytes to server
+
+    let password_file = ServerRegistration::<Default>::try_from(password_file_bytes).unwrap();
+    let mut server_rng = OsRng;
+    let (l2, server_state) = ServerLogin::start(
+        password_file,
+        &server_kp.private(),
+        LoginFirstMessage::try_from(&l1_bytes[..]).unwrap(),
+        &mut server_rng,
+    )
+    .unwrap();
+    let l2_bytes = l2.to_bytes();
+
+    // Server sends l2_bytes to client
+
+    let result = client_state.finish(
+        LoginSecondMessage::try_from(&l2_bytes[..]).unwrap(),
+        &server_kp.public(),
+        &mut client_rng,
+    );
+
+    if result.is_err() {
+        // Note that at this point, the client knows whether or not the login
```

Comment:
Since the only sequencing bit in this example is in the `main()`, would it be possible to instead place the description of this logic there?

Reply:
Good point!